### PR TITLE
Migrate NMState to use kubernetes-nmstate-operator if available

### PR DESCRIPTION
**Pull request by bcrochet**
This patch make CNAO use the kubernetes-nmstate-operator if it is
running and available. The only prereq for this is to have the
operator pod running. If that's detected via the
app=kubernetes-nmstate-operator label on a Deployment, then CNAO
will only render an nmstates.nmstate.io CR to kick off the handler
and webhook from the operator.

If the operator is added, it also has upgrade logic to move to using
the operator instead of installing kubernetes-nmstate directly. It
will remove the old instance.

Signed-off-by: Brad P. Crochet <brad@redhat.com>

**What this PR does / why we need it**:
See above.

**Special notes for your reviewer**:
See https://nmstate.io/kubernetes-nmstate/deployment.html for installing the operator

Depends on https://github.com/kubevirt/cluster-network-addons-operator/pull/828

**Release note**:

```release-note
Migrate Kubernetes NMState install to use kubernetes-nmstate-operator if it is already available in the cluster. This will work on upgrade as well.
```

### CODE REVIEW

**Comment by qinqon** [comment]
@bcrochet Let's fix CI before review this

File: pkg/network/nmstate.go
Comment (qinqon) [inline pkg/network/nmstate.go:133]: We have bump CNAO to latest nmstate and now there is a new Deployment "cert-manager" it has to be added here.

**Review by qinqon — CHANGES_REQUESTED** [review]
firt pass, let's add cert-manager too.

File: pkg/network/nmstate.go
Comment (bcrochet) [reply pkg/network/nmstate.go]: Done.

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go
Comment (bcrochet) [reply pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:541]: > Do we have to check if the NMState CR CNAO has is compatible with the installed CRD from the operator ?
> 

I don't think so. Versioning of the NMState API should take care of that. And the nmstate operator should handle that.

> Imaging that we are installing CNAO on openshift with old NMState operator we will apply a possible too new NMState CR and the opposite CNAO being old and openshift NMState operator newer.

So, an OpenShift cluster already has NMState operator installed.. CNAO would try to install a newer CR? That's an interesting scenario, but again, I think versioning would come into play there. I suppose we could detect in CNAO that the version we are trying to install is what we are trying to install... As for he CNAO being old and nmstate operator being newer... that should be handled by the nmstate operator.

> 
> Also in case the CR has being already applied I would do nothing to reduce possible problems.

If the CR is already applied, it should be a big no-op. But it would be interesting to find out what happens there.

**Comment by qinqon** [comment]
/hold

We have to bump nmstate at CNAO before merging this since it depends on the fix for handler lock

**Comment by qinqon** [comment]
/hold cancel

Fix mertged and released

**Comment by qinqon** [comment]
@bcrochet we hav already a handler lock test on kubernetes-nmstate so the function is working fine, can we do at least a manual test with this PR for CNAO to everify that it's working, it would be even better a e2e tests at CNAO but don't know if it's worth it.

**Comment by bcrochet** [comment]
> @bcrochet we hav already a handler lock test on kubernetes-nmstate so the function is working fine, can we do at least a manual test with this PR for CNAO to everify that it's working, it would be even better a e2e tests at CNAO but don't know if it's worth it.

Quick test shows that it does indeed keep the operator-installed instance (outside of CNAO) from taking the lock.

**Comment by qinqon** [comment]
@bcrochet it's worth adding some e2e test here ?

**Comment by bcrochet** [comment]
> @bcrochet it's worth adding some e2e test here ?

A couple would be a good idea.

File: test/e2e/workflow/deployment_test.go
Comment (qinqon) [inline test/e2e/workflow/deployment_test.go:404]: Can we apply the manifest from the kubernetes-nmstate version at components.yaml ?

**Review by qinqon — CHANGES_REQUESTED** [review]
Let's take the manifests from components.yaml

File: test/manifests/kubernetes-nmstate-operator.yaml
Comment (qinqon) [inline test/manifests/kubernetes-nmstate-operator.yaml:3]: we don't need this anymore right ?

File: pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go
Comment (qinqon) [inline pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:544]: Don't we have to specify the namespace ?

Comment (bcrochet) [reply pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:544]: We wouldn't necessarily know what the namespace is.

**Pull request merged**
